lib/backend/etcdv3/etcdv3.go

```diff
@@ -0,0 +1,465 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdv3
+
+import (
+	"context"
+	goerrors "errors"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"github.com/coreos/etcd/pkg/transport"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+)
+
+var (
+	clientTimeout = 30 * time.Second
+)
+
+type EtcdV3Client struct {
+	etcdClient *clientv3.Client
+}
+
+func NewEtcdV3Client(config *apiconfig.EtcdConfig) (api.Client, error) {
+	// Determine the location from the authority or the endpoints.  The endpoints
+	// takes precedence if both are specified.
+	etcdLocation := []string{}
+	if config.EtcdEndpoints != "" {
+		etcdLocation = strings.Split(config.EtcdEndpoints, ",")
+	}
+
+	if len(etcdLocation) == 0 {
+		log.Info("No etcd endpoints specified in etcdv3 API config")
+		return nil, goerrors.New("no etcd endpoints specified")
+	}
+
+	// Create the etcd client
+	tlsInfo := &transport.TLSInfo{
+		CAFile:   config.EtcdCACertFile,
+		CertFile: config.EtcdCertFile,
+		KeyFile:  config.EtcdKeyFile,
+	}
+	tls, _ := tlsInfo.ClientConfig()
+
+	cfg := clientv3.Config{
+		Endpoints:   etcdLocation,
+		TLS:         tls,
+		DialTimeout: clientTimeout,
+	}
+
+	// Plumb through the username and password if both are configured.
+	if config.EtcdUsername != "" && config.EtcdPassword != "" {
+		cfg.Username = config.EtcdUsername
+		cfg.Password = config.EtcdPassword
+	}
+
+	client, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EtcdV3Client{etcdClient: client}, nil
+}
+
+// Create an entry in the datastore.  If the entry already exists, this will return
+// an ErrorResourceAlreadyExists error and the current entry.
+func (c *EtcdV3Client) Create(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"model-key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	logCxt.Debug("Processing Create request")
+	key, value, err := getKeyValueStrings(d)
+	if err != nil {
+		return nil, err
+	}
+	logCxt = logCxt.WithField("etcdv3-key", key)
+
+	putOpts, err := c.getTTLOption(d)
+	if err != nil {
+		return nil, err
+	}
+
+	// Checking for 0 version of the key, which means it doesn't exists yet,
+	// and if it does, get the current value.
+	logCxt.Debug("Performing etcdv3 transaction for Create request")
+	txnResp, err := c.etcdClient.Txn(context.Background()).If(
+		clientv3.Compare(clientv3.Version(key), "=", 0),
+	).Then(
+		clientv3.OpPut(key, value, putOpts...),
+	).Else(
+		clientv3.OpGet(key),
+	).Commit()
```

Comment:
so interestingly, this is how they're doing it in k8s API server: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L175-L179

Reply:
yes - i think i stole the idea from there.

Reply:
Actually this was lifted from the etcdv3 PR which I've modified a little.  So no from the k8s code.

So it does need a decent review.  That said, G5 has already given it a fairly in-depth review.